Feat: Partial repayments

PowVT opened the pull request:
This PR includes updates to the RC, OC, and LC contracts to support multiple repayments. 
- There is no penalty on the borrower for repaying or rolling over early on in the loan lifecycle.
- The minimum repayment amount at any time in the loan is equivalent to the minimum interest due at that time. If the amount specified by the borrower is over the interest due, the rest of the repayment goes towards the loans balance. When the loans balance goes to zero, a loan is marked as repaid and the collateral is released to the borrower from escrow.

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -224,28 +224,46 @@ contract LoanCore is
+        // if loan is completely repaid burn both notes
+        if (loanRepaid) _burnLoanNotes(loanId);
+
+        // collect prinicpal and interest from borrower
```
Typo here

PowVT replied:
I had my spell checker off 😢. TY, fixed!

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -224,28 +224,46 @@ contract LoanCore is
 
-        // Send collected principal + interest, less fees, to lender
+        // if loan is completely repaid burn both notes
+        if (loanRepaid) _burnLoanNotes(loanId);
```
Should go down on line 260 next to collateral transfer? We should handle all effects of `loanRepaid == true` in the same place IMO

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -258,32 +276,52 @@ contract LoanCore is
+        // DO NOT burn LenderNote until receipt is redeemed
         address borrower = borrowerNote.ownerOf(loanId);
-        borrowerNote.burn(loanId);
+        if (loanRepaid) borrowerNote.burn(loanId);
```
Same as above re: grouping the `loanRepaid == true` effects

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -258,32 +276,52 @@ contract LoanCore is
+            emit LoanRepaid(loanId);
+        } else {
+            // if loan is not completely repaid emit payment event
+            emit LoanPayment(loanId);
```
I would emit `LoanPayment` in both scenarios (so on closure, we emit 2 events in this fn)

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -368,12 +407,20 @@ contract LoanCore is
-        lenderNote.burn(loanId);
+
+        // if the loan has been completely repaid and no more repayments are expected
+        if (loans[loanId].balance == 0) {
```
Seems more intuitive to read `LoanState` here? Or if we want to use `balance == 0` as truth, can we remove the `LoanState` enum altogether and save some gas?

PowVT replied:
No I think checking LoanState is the best indicator here. Made this change

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -389,34 +436,39 @@ contract LoanCore is
 
         // State change for old loan
         data.state = LoanLibrary.LoanState.Repaid;
+        data.balance = 0;
```
Same thing here, just looks like cleanup to make sure rollovers handle proration cleanly right? Not related to partial repayments

PowVT replied:
My intention with these state updates on 469 and 470 were to reflect the rollover 'repayment' in the old loanData to show that it was paid off by the rollover action. I wasnt thinking about the zero balance checks and as mentioned in last comment think it is best to use the LoanState.

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -743,12 +800,20 @@ contract LoanCore is
-        loans[loanId].interestAmountPaid = _amountFromPayer - data.balance;
-        loans[loanId].balance = 0;
+        // state changes
+        loanRepaid = false;
```
Can save an assignment here by just having `loanRepaid = (loans[loanId].state == LoanLibrary.LoanState.Repaid)` after the `if` block.

Or, don't return `LoanRepaid` and just have callers of _handleRepay read the `state` from `LoanData` (can ignore this if we remove that enum altogether)

PowVT replied:
Yes good point, I removed the local var and changed to the calling functions looking at the LoanState enum

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -743,12 +800,20 @@ contract LoanCore is
+        }
+
+        loans[loanId].interestAmountPaid += _interestAmount;
+        loans[loanId].balance -= _paymentToPrincipal;
```
Revert check for possible underflow here? From LoanCore's perspective, `_paymentToPrincipal` is untrusted input and could be any value, plus it seems like underflow is very possible (since we are handling the "**greater than** or equal to" scenario on line 805 above)

PowVT replied:
Yes, no checked from LC's perspective. I added this line:
` if (_paymentToPrincipal > data.balance) revert LC_ExceedsBalance(_paymentToPrincipal, data.balance);`
and change the >= to be == in the check for if this is the final repayment.

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -714,22 +766,27 @@ contract LoanCore is
+        uint256 _amountToLender,
+        uint256 _interestAmount,
+        uint256 _paymentToPrincipal
+    ) internal returns (LoanLibrary.LoanData memory data, bool loanRepaid) {
```
What happens if `_interestAmount + _paymentToPrincipal != _amountFromPayer`? Related question, if we have the latter two values, what do we also need `_amountFromPayer` for?

PowVT replied:
Yes correct, I made changes such that the `_amountFromPayer` is not needed and updated LC tests.

kkennis commented on contracts/RepaymentController.sol:
```diff
@@ -134,10 +156,8 @@ contract RepaymentController is IRepaymentController, InterestCalculator, FeeLoo
     function redeemNote(uint256 loanId, address to) external override {
```
The `NoteRedeemed` event probably needs a new `bool` field to tell whether the note has been burned

PowVT replied:
The burning of the lenderNote will emit an event in this case. I didnt want to duplicate the logic in the event as well. Going to get with @sandbochs to see what makes the most sense for him from a FE view

PowVT replied:
nvm, OOO till the 22nd. Going to check with @luckyOdisetti

kkennis replied:
Ah true, I think that burn probably covers being able to differentiate. I'm cool w/ either implementation in that case.

PowVT commented on contracts/OriginationController.sol:
```diff
@@ -953,13 +953,15 @@ contract OriginationController is
 
         loanId = loanCore.rollover(
```
The only change I made here was to update the `interestAmountPaid` state in LC's LoanData. I figure it is best to keep the loanData of the old loan to be accurate to the interest that was repaid during the rollover.

PowVT commented on contracts/interfaces/IRepaymentController.sol:
```diff
@@ -5,9 +5,9 @@ pragma solidity 0.8.18;
     // ============== Lifeycle Operations ==============
 
-    function repay(uint256 loanId) external;
+    function repay(uint256 loanId, uint256 amount) external;
```
I added explicit routes to the RC for 'full' repayments. Even though yes correct about the max unit256, it seems like an anti pattern and could save a little gas since the amount checks are not needed in full repayment. Added tests for this change to the `PartialRepayments.ts` file.

For tests showing repayments where the borrower specifies and amount that is larger than the amount due, see the RepaymentController.ts and Rollover.ts files

kkennis commented on contracts/LoanCore.sol:
```diff
@@ -743,12 +798,20 @@ contract LoanCore is
+            // If the payment is equal to the balance, the loan is repaid.
+            // Full repayments set the _paymentToPrincipal to the balance.
+            // For partial repayments where the amount to pay is greater than the balance,
+            // the _paymentToPrincipal is set to the balance.
```
Not sure I follow this comment. The RepaymentController does this, right? But LoanCore is "dumb" to how the RepaymentController works. We already have the greater-than check above, but this seems to be saying that LoanCore is doing something that it is not actually doing.

PowVT replied:
Yes, makes sense. I removed the last three lines of that comment.

PowVT commented on contracts/LoanCore.sol:
```diff
@@ -767,30 +764,31 @@ contract LoanCore is
         // Ensure valid initial loan state when repaying loan
         if (data.state != LoanLibrary.LoanState.Active) revert LC_InvalidState(data.state);
 
+        totalRepayment = _paymentToPrincipal + _interestAmount;
```
Since the sum of `_paymentToPrincipal + _interestAmount` is used in a few places in the `_handleRepay` fn and also the `repay` and `forceRepay` fn's, I thought it was cleaner to do the math here and reuse the result in those other places.

kkennis replied:
But you already get this same value on line 778, right?

PowVT commented on contracts/LoanCore.sol:
```diff
@@ -411,7 +408,7 @@ contract LoanCore is
 
         // if the loan has been completely repaid and no more repayments are expected
-        if (loans[loanId].balance == 0) {
+        if (loans[loanId].state == LoanLibrary.LoanState.Repaid) {
```
Yes, this scenario would mean the lender couldn't receive those partial repayments.

For a fix, I think we could move away from having the holder of the lenderNote be able to claim and instead track for repayment claims with a mapping like this `noteReceipts[loanId][lender]`. This way even if a lender transfers the note, they can still collect payment from before they transfer it in the event it is burned/sold/transferred. What do you think?

With this, and the other prorated changes, we should update the token images to reflect the new LoanData params like interest paid, balance, last payment time, etc.. In order to provide a better experience when selling the Borrower/LenderNotes

kkennis replied:
Definitely agree on the token images.

I think it's a matter of taste and user protection; I don't think we should move away from the LenderNote, that increases complexity. Right now we can say that the holder of the note holds all rights and obligations associated with the lending side, but if we move away, that would no longer be true, and we'd need to start putting "except..." in a lot of places.

I think it's a matter of taste here, should we protect users or not? If so, maybe we can just add a revert check in `claim` if there are unclaimed funds associated with the note? Or pay them out automatically on claim as well?

PowVT replied:
I am a fan of adding the revert statement saying the lender should collect their payment before claiming the asset. This seems like the most straightforward approach. We could combine the `redeemNote` logic in the claim flow, but I think it would bloat the regular liquidation flow where payments are sent directly to the lender. Best to keep the happy path as efficient as possible imo

kkennis replied:
Yep, agreed - think that sounds good to me!

PowVT commented on contracts/RepaymentController.sol:
```diff
@@ -223,7 +262,56 @@ contract RepaymentController is IRepaymentController, InterestCalculator, FeeLoo
+     * @return interestAmount       The amount of interest due.
+     * @return paymentToPrincipal   The portion of the repayment amount that goes to principal.
+     */
+    function _prepareRepayFull(uint256 loanId)
```
Yes good point, I added an if statement to the `_prepareRepay` fn where is says in the amount is MAX_UINT, set the `paymentToPrincipal` to the `data.balance` else preform amount checks (like before).

kkennis commented on contracts/RepaymentController.sol:
```diff
@@ -152,30 +192,62 @@ contract RepaymentController is IRepaymentController, InterestCalculator, FeeLoo
-        uint256 principalFee = (terms.principal * data.feeSnapshot.lenderPrincipalFee) / BASIS_POINTS_DENOMINATOR;
-
-        amountFromBorrower = terms.principal + interest;
+        if(amount == MAX_INT) {
```
Can remove this (along with the defined constant). Won't lines 238-240 already handle this cleanly?

kkennis reviewed (commented):
Everything looks great, but seeing some missing tests:

- Tests that call `repayFull`
- Tests that trigger `LC_ExceedsBalance`

I also see that `PartialRepayments.ts` is a bit flaky.....it's probably extremely difficult to get the exact timing right. When I ran that test file 10 times it failed about half of the time. I think it would be acceptable to have a range, sort of a tolerance range to prevent needing to fight the way hardhat times transactions (and possible CI issues).

You could use something like [chai-almost](https://www.chaijs.com/plugins/chai-almost/) or just manual checks, e.g.:

```
            // Expecting 1500
            expect(effectiveInterestRate).to.be.gt(1490).and.lt(1510)
```

I do think it's important that we get the tests pass every time, so I'd rather add some tolerance to the values as opposed to shooting for the exact value every time but knowing Hardhat might not cooperate

Mouzayan reviewed (commented):
Nicely done!
See below a question and a couple of nits.

PowVT commented on contracts/RepaymentController.sol:
```diff
@@ -63,33 +66,75 @@ contract RepaymentController is IRepaymentController, InterestCalculator, FeeLoo
-        // call repay function in loan core -  msg.sender will pay the amountFromBorrower
-        loanCore.repay(loanId, msg.sender, amountFromBorrower, amountToLender);
+    function repay(uint256 loanId, uint256 amount) external override {
+        // no zero amount check, minimum principal could be 0 for this payable currency
```
Hey, good q. There is a scenario in the tests where a loan could be started with a zero amount principal. This can potentially occur because in the `setAllowedPayableCurrencies` fn there is no ZeroAmount check on the `currencyData[i].minPrincipal` input. So it is possible to have 0 principal loans, even none of the currently whitelisted currencies in V3 support this because there minimums are higher than zero, it is a possibility. The test for this can be found in the RepaymentController.ts file titled "repays a loan with 0 amount due"

PowVT commented on test/PartialRepayments.ts:
```diff
@@ -0,0 +1,1605 @@
+
+            // total repayment amount
+            const total = ethers.utils.parseEther("105");
+            const repayAdditionalAmount = total.sub(await mockERC20.balanceOf(borrower.address));
```
For tests where we do not pre-calculate the interest amount, this pattern of minting the 'additionalAmount' works well when we need to check the balance of the borrower after the tx happens. Since the borrower receives the principal when the loan is started, minting the total logically doesnt make sense, where are minting the difference does so that our borrower balances are exact.

 It is also used in places where there are origination fees so you dont have to calculate the fees it will mint the difference to the borrower and the expect statements later are more straight forward.

PowVT commented:
Made updates based on the last round of reviews, adding tolerances to the tests works well. Also added the `// expecting: value x` statements anywhere the expect statements are written with tolerances. Also removed the shortcut for MAX_UINT in `repayFull`. 

I did find a small bug in the `getCloseEffectiveInterestRate` fn where it did not calculate the proper value when the `lastAccrualTimestamp` was greater than the loan duration so I added in a if statement to handle this. Also added a few expect statements to the PartialRepayments.ts test checking this.
